pkg/services/ngalert/schedule/alert_rule.go

```diff
@@ -68,3 +80,272 @@ func (a *alertRuleInfo) update(lastVersion ruleVersionAndPauseStatus) bool {
 func (a *alertRuleInfo) stop(reason error) {
 	a.stopFn(reason)
 }
+
+//nolint:gocyclo
+func (ruleInfo *alertRuleInfo) ruleRoutine(key ngmodels.AlertRuleKey, sch *schedule) error {
```

Comment:
parameters are swapped here - `ruleInfo *alertRuleInfo` and `sch *schedule`.
no further changes to keep the diff sane. names are kept the same for the same reason.

rest of the code, and dependent functions are 100% untouched, copy/pasted from `schedule.go`.

Reply:
this obviously isn't the cleanest (bad naming, the entire scheduler being passed in is weird) but it's no less knowledge than before. will be improved in future PRs, just couldn't do it here because the code moving breaks the diff.

Reply:
the linter forced me to fix the name :skull: